Allow `[p]ban` to hackban and rename `[p]hackban` to `[p]massban`

phenom4n4n opened the pull request:
### Type

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes
Allows the `ban` command to hackban users which aren't in the guild, with support for replacing tempbans. Additionally renames `hackban` to `massban`, but keeps `hackban` as an alias. 
The rest of the changes were what got hit by style reformatting. 
Fixes #4419

Commit:
ban revamp

Jackenmen commented:
It seems like your commit is failing the style check, please make sure that you are using up-to-date version of black (20.8b1)

Commit:
black & converters fix

phenom4n4n commented:
I updated black and switched the converters to include discord.User and RawUserIds.

Jackenmen commented:
@phenom4n4n Thanks, looks a lot better now. Could you also manually revert the change to `redbot/cogs/admin/admin.py` as it seems like double reformatting caused it to misplace the comments there.
Feel free to do it later, when the actual full review comes up though, it's not like this breaks anything functionally.

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -260,29 +291,32 @@ async def ban(
         guild = ctx.guild
         if days is None:
             days = await self.config.guild(guild).default_days()
+        if isinstance(user, int):
+            try:
+                user = await self.bot.fetch_user(user)
+            except discord.errors.NotFound:
+                await ctx.send_help()
+                return
```
This method should not try to fetch a user on ban as fetch user rate limit is low. It should (similarly to `[p]hackban`) use `discord.Object`s for banning when we only have a raw user ID.

Jackenmen commented on redbot/cogs/admin/admin.py:
```diff
@@ -76,8 +76,8 @@ def __init__(self):
         self.config.register_guild(
-            announce_channel=None,  # Integer ID
-            selfroles=[],  # List of integer ID's
+            announce_channel=None,
+            selfroles=[],  # Integer ID  # List of integer ID's
```
This should be changed back to how it was.
```suggestion
            announce_channel=None,  # Integer ID
            selfroles=[],  # List of integer ID's
```

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -70,48 +70,79 @@ async def ban_user(
+
+            ban_type = "hackban"
+
+        if not (0 <= days <= 7):
+            return _("Invalid days. Must be between 0 and 7.")
```
This should happen at the beginning of the function, because this condition needs to be fulfilled in order for the bot to do any actions.

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -246,7 +277,7 @@ async def kick(self, ctx: commands.Context, user: discord.Member, *, reason: str
         self,
         ctx: commands.Context,
-        user: discord.Member,
+        user: Union[discord.Member, discord.User, RawUserIds],
```
See the comment below for more information.
```suggestion
        user: Union[discord.Member, RawUserIds],
```

Jackenmen replied:
As explained in the server, `discord.User` converter here as well as handling for it in `ban_user()` should be removed.

Jackenmen reviewed (changes requested):
This is probably not a full review, but I posted review comments on what I noticed.
I'll do a full review once this gets modified in order to work without use of `discord.User` objects.

phenom4n4n committed:
Update redbot/cogs/admin/admin.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

Commit:
discord.object

Commit:
remove discord.user converter

Commit:
black

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -70,48 +70,81 @@ async def ban_user(
         except discord.Forbidden:
             return _("I'm not allowed to do that.")
+        except discord.NotFound:
+            return _(f'User "{user.id}" not found')
```
f-strings can't be used with `_()` as the string in `_()` can't be dynamic and needs to be formatted afterward:
```suggestion
            return _('User with ID {user_id} not found').format(user_id=user.id)
```

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -70,48 +70,81 @@ async def ban_user(
+                    )
                 )
-            )
+                end_result = _(f"{user} was upgraded from a temporary to a permanent ban.")
```
f-strings can't be used with `_()` as the string in `_()` can't be dynamic and needs to be formatted afterward:
```suggestion
                end_result = _(
                    "User with ID {user_id} was upgraded from a temporary to a permanent ban."
                ).format(user_id=user.id)
```

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -70,48 +70,81 @@ async def ban_user(
-            log.info(
-                "{}({}) banned {}({}), deleting {} days worth of messages.".format(
-                    author.name, author.id, user.name, user.id, str(days)
+            if removed_temp == True:
```
We mostly adhere to PEP 8 style guide, so this should just be using `if greeting:` without `==`.
````suggestion
            if removed_temp:
```

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -70,48 +70,81 @@ async def ban_user(
+                        author.name, author.id, ban_type, user.name, user.id, str(days)
+                    )
+                )
+                end_result = _("Done. It was about time.")
```
Having this new behavior instead of the old `True` breaks `[p]massban`'s usage of this method.

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -246,7 +279,7 @@ async def kick(self, ctx: commands.Context, user: discord.Member, *, reason: str
     async def ban(
```
As proposed earlier in Discord by Flame, it would be good if `ban`'s help docstring said that you need to use user ID/mention if the user isn't in the server.

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -70,48 +70,81 @@ async def ban_user(
 
         queue_entry = (guild.id, user.id)
         try:
             await guild.ban(user, reason=audit_reason, delete_message_days=days)
```
This should only be done if `removed_temp` is `False` as otherwise the user is already banned and we just had to upgrade the tempban to permban in the bot's config.

Jackenmen commented on redbot/cogs/mod/kickban.py:
```diff
@@ -396,9 +428,7 @@ def remove_processed(ids):
-                        errors[user_id] = _("User {user_id} does not exist.").format(
-                            user_id=user_id
-                        )
+                        errors[user_id] = _('User "{user_id} not found').format(user_id=user_id)
```
A bit clearer:
```suggestion
                        errors[user_id] = _(
                            'User with ID "{user_id}" not found'
                        ).format(user_id=user_id)
```

Jackenmen reviewed (changes requested):
Full code review

phenom4n4n committed:
Update redbot/cogs/mod/kickban.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
Update redbot/cogs/mod/kickban.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
Update redbot/cogs/mod/kickban.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
Update redbot/cogs/mod/kickban.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

Commit:
.

Commit:
incomplete

Commit:
massban support

Commit:
black

Jackenmen committed:
Use 2-tuple to separate result and the message in `ban_user()`

This also fixes the issue with `True` being equal to `1` which caused a problem with previously returned types

Jackenmen committed:
Whoops...

Jackenmen committed:
trailing whitespace...

Jackenmen committed:
I missed this one

Jackenmen committed:
Update kickban.py

Jackenmen committed:
Update kickban.py

Jackenmen reviewed (approved):
LGTM, thanks for all the work :)
